[Grid][Cleanup] Adopt RenderChildIterator throughout grid code

nullhook opened the pull request:
#### f0515d83ab79099506f4d7ebe9140bca98f52111
<pre>
[Grid][Cleanup] Adopt RenderChildIterator throughout grid code
<a href="https://rdar.apple.com/146779702">rdar://146779702</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=289555">https://bugs.webkit.org/show_bug.cgi?id=289555</a>

Reviewed by Alan Baradlay.

Using RenderChildIterator to utilize range-based for loop

* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::populateExplicitGridAndOrderIterator):
(WebCore::RenderGrid::layoutGridItems):
(WebCore::RenderGrid::layoutMasonryItems):

Canonical link: <a href="https://commits.webkit.org/292043@main">https://commits.webkit.org/292043@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1a6aead901cee63a487ac7966647d45915057b4a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/94734 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/14323 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/4149 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/99753 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/45226 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/96784 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/14601 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/22743 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/72271 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/29579 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/97736 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/10887 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/85530 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/52602 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/10580 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/3240 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/44564 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/80799 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/3339 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/101795 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/21762 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/15888 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/81266 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/22012 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/81552 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/80643 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20151 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/25204 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/2602 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/14997 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/21739 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/26854 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/21402 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/24873 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/23139 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/rendering/RenderGrid.cpp:
```diff
@@ -1106,8 +1106,8 @@ Vector<LayoutRect> RenderGrid::gridItemsLayoutRects()
-    for (RenderBox* gridItem = currentGrid().orderIterator().first(); gridItem; gridItem = currentGrid().orderIterator().next())
-        items.append(gridItem->frameRect());
+    for (RenderBox& gridItem : childrenOfType<RenderBox>(*this))
+        items.append(gridItem.frameRect());
```
We don't care about order anymore?

nullhook replied:
good catch, I'm revering this specific change so it maintains the order. however, the function of this logic isn't referenced anywhere.

sammygill commented on Source/WebCore/rendering/RenderGrid.cpp:
```diff
@@ -1087,11 +1087,11 @@ void RenderGrid::placeItemsOnGrid(std::optional<LayoutUnit> availableLogicalWidt
 #if ASSERT_ENABLED
-    for (auto* gridItem = currentGrid().orderIterator().first(); gridItem; gridItem = currentGrid().orderIterator().next()) {
-        if (currentGrid().orderIterator().shouldSkipChild(*gridItem))
+    for (RenderBox& gridItem : childrenOfType<RenderBox>(*this)) {
```
I think we try to use auto throughout (newly written) layout code so maybe we should be consistent

nullhook replied:
changed

sammygill reviewed (commented):
The format of the commit message should probably be cleaned up a bit since there's some redundancy from trying to squash multiple of them together

sammygill commented on Source/WebCore/rendering/RenderGrid.cpp:
```diff
@@ -2497,8 +2497,8 @@ unsigned RenderGrid::numTracks(GridTrackSizingDirection direction) const
     ASSERT(!currentGrid().needsItemsPlacement());
-    for (RenderBox* gridItem = currentGrid().orderIterator().first(); gridItem; gridItem = currentGrid().orderIterator().next())
-        paintChild(*gridItem, paintInfo, paintOffset, forChild, usePrintRect, PaintAsInlineBlock);
+    for (RenderBox& gridItem : childrenOfType<RenderBox>(*this))
```
Looks like we are painting the children in a different order from before. It's not super clear to me if this is ok

alanbaradlay replied:
definitely not ok. please do not change behavior in a cleanup PR.

nullhook replied:
i have reverted this change

sammygill commented on Source/WebCore/rendering/RenderGrid.cpp:
```diff
@@ -1464,43 +1464,43 @@ void RenderGrid::layoutMasonryItems(GridLayoutState& gridLayoutState)
-        if (currentGrid().orderIterator().shouldSkipChild(*gridItem)) {
-            if (gridItem->isOutOfFlowPositioned())
-                prepareGridItemForPositionedLayout(*gridItem);
+    for (RenderBox& gridItem : childrenOfType<RenderBox>(*this)) {
```
Should probably be auto for consistency

sammygill commented on Source/WebCore/rendering/RenderGrid.cpp:
```diff
@@ -1419,43 +1419,43 @@ void RenderGrid::layoutGridItems(GridLayoutState& gridLayoutState)
-        if (currentGrid().orderIterator().shouldSkipChild(*gridItem)) {
-            if (gridItem->isOutOfFlowPositioned())
-                prepareGridItemForPositionedLayout(*gridItem);
+    for (RenderBox& gridItem : childrenOfType<RenderBox>(*this)) {
```
Should probably be auto for consistency

nullhook replied:
changed to `auto`

sammygill commented on Source/WebCore/rendering/RenderGrid.cpp:
```diff
@@ -1149,35 +1149,35 @@ void RenderGrid::populateExplicitGridAndOrderIterator()
 
-    for (RenderBox* gridItem = firstChildBox(); gridItem; gridItem = gridItem->nextSiblingBox()) {
-        if (!populator.collectChild(*gridItem))
+    for (RenderBox& gridItem : childrenOfType<RenderBox>(*this)) {
```
Should probably be auto for consistency

nullhook replied:
ditto

sammygill commented on Source/WebCore/rendering/RenderGrid.cpp:
```diff
@@ -1087,11 +1087,11 @@ void RenderGrid::placeItemsOnGrid(std::optional<LayoutUnit> availableLogicalWidt
 #if ASSERT_ENABLED
-    for (auto* gridItem = currentGrid().orderIterator().first(); gridItem; gridItem = currentGrid().orderIterator().next()) {
-        if (currentGrid().orderIterator().shouldSkipChild(*gridItem))
+    for (auto& gridItem : childrenOfType<RenderBox>(*this)) {
```
This *might* be ok since it looks like this is just some debug logic to sanity check grid item placement occurred properly, which should be independent from order, but it probably deserves a justification/mention in the commit message if we want to keep it

alanbaradlay replied:
> This _might_ be ok since it looks like this is just some debug logic to sanity check grid item placement occurred properly, which should be independent from order, but it probably deserves a justification/mention in the commit message if we want to keep it

While this might be ok, cleanup PRs are not supposed to change behavior.

nullhook replied:
makes sense. i have reverted the change on sites where ordering matters

nullhook committed:
[Grid][Cleanup] Adopt RenderChildIterator throughout grid code
rdar://146779702
https://bugs.webkit.org/show_bug.cgi?id=289555

Reviewed by Alan Baradlay.

Using RenderChildIterator to utilize range-based for loop

* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::populateExplicitGridAndOrderIterator):
(WebCore::RenderGrid::layoutGridItems):
(WebCore::RenderGrid::layoutMasonryItems):

Canonical link: https://commits.webkit.org/292043@main
